contributing/building-a-bento-amp-extension.md

```diff
@@ -0,0 +1,681 @@
+-   **Example Usage**: amp-accordion, amp-selector
+
+## Element styling
+
+You can write a stylesheet to style your element to provide a minimal
+visual appeal. Your element structure should account for whether you
+want users (publishers and developers using your element) to customize
+the default styling you're providing and allow for easy CSS classes
+and/or well-structure DOM elements.
+
+Element styles are loaded when the element script itself is included in
+an AMP doc. You tell AMP which CSS or JSS belongs to this element when
+registering the element (see below).
+
+## Register element with AMP
+
+Once you have implemented your AMP element, you need to register it with
+AMP; all AMP extensions are prefixed with `amp-`.
+
+One way to specify the appropriate styles is to tell AMP which class to use for this tag name and which CSS to load.
+
+```javascript
+import {CSS} from '../../../build/amp-my-element-0.1.css';
+
+AMP.extension('amp-my-element', '0.1', (AMP) => {
+  AMP.registerElement('amp-my-element', AmpMyElement, CSS);
+});
+```
+
+Alternatively, you may have defined styles via JSS that are compiled into CSS and attached via the shadow stylesheet. To tell AMP which class to use, override the `shadowCss` property:
+
+```javascript
+import {CSS} from './my-element.jss';
+
+/** @override */
+AmpMyElement['shadowCss'] = CSS;
+```
+
+## Actions and events
+
+AMP provides a framework for [elements to fire their own
+events](https://github.com/ampproject/amphtml/blob/master/spec/amp-actions-and-events.md)
+to allow users of that element to listen and react to the events. For
+example, the `amp-base-carousel` extension fires a `slideChange` event.
+This allow publishers to listen to that event and react to it, for example, by updating an `amp-selector` state to match the current slide being shown.
+
+The other part of the event-system in AMP is actions. When listening to
+an event on an element usually you'd like to trigger an action (possibly
+on other elements). For example, in the example above, the publisher is
+executing the `toggle` action on `amp-selector`.
+
+The syntax for using this on elements is as follow:
+
+```html
+<amp-base-carousel
+  on="slideChange:my-selector.toggle(index=e.index, value=true)"
+></amp-base-carousel>
+```
+
+To fire events on your element use AMP's action service and the
+`.trigger` method.
+
+```javascript
+actionServiceForDoc(doc.documentElement).trigger(
+  this.element_,
+  'slideChange',
+  createCustomEvent(
+    win,
+    `amp-base-carousel.${name}`,
+    dict({'index': index})
+  ),
+  ActionTrust.DEFAULT
+);
+```
+
+And to expose actions use `registerApiAction` method that your element
+inherits from `PreactBaseElement`. Note, this should correspond directly with the API exposed in the corresponding Preact compnent via `useImperativeHandle`, and the component should be defined using a `forwardRef` accordingly.
```

Comment:
I think we should link `useImperativeHandle` and `forwardRef` to React or Preact docs. Might be better for Preact since the APIs we normally test against.

Reply:
I'm not finding any links when searching "useImperativeHandle site:preactjs.com", so I added the React documentation there. For forwardRef it is mentioned in a "Switching to Preact" document, so I've added a link to that: https://preactjs.com/guide/v10/switching-to-preact/#forwardref 

LMK if these are suitable or if you had others in mind.

Reply:
LG